src/runner/index.js

```diff
@@ -204,83 +199,150 @@ export default class Runner extends EventEmitter {
     }
 
+    static _isStreamMock (obj) {
+        return obj &&
+               typeof obj.write === 'function' &&
+               typeof obj.end === 'function';
+    }
+
+    static _validateReporterOutput (obj) {
+        if (obj !== void 0 &&
+            typeof obj !== 'string' &&
+            !isWritableStream(obj) &&
+            !Runner._isStreamMock(obj))
+            throw new GeneralError(MESSAGE.invalidReporterOutput);
+    }
+
+    _setBootstrapperOptions () {
+        this.bootstrapper.sources                     = this.configuration.getOption(OPTION_NAMES.src) || this.bootstrapper.sources;
+        this.bootstrapper.browsers                    = this.configuration.getOption(OPTION_NAMES.browsers) || this.bootstrapper.browsers;
+        this.bootstrapper.concurrency                 = this.configuration.getOption(OPTION_NAMES.concurrency) || this.bootstrapper.concurrency;
+        this.bootstrapper.appCommand                  = this.configuration.getOption(OPTION_NAMES.appCommand) || this.bootstrapper.appCommand;
+        this.bootstrapper.appInitDelay                = this.configuration.getOption(OPTION_NAMES.appInitDelay) || this.bootstrapper.appInitDelay;
+        this.bootstrapper.disableTestSyntaxValidation = this.configuration.getOption(OPTION_NAMES.disableTestSyntaxValidation);
+        this.bootstrapper.filter                      = this.configuration.getOption(OPTION_NAMES.filter) || this.bootstrapper.filter;
+        this.bootstrapper.reporters                   = this.configuration.getOption(OPTION_NAMES.reporter) || this.bootstrapper.reporters;
+    }
+
+    _prepareReporters (name, outStream) {
+        let reporters = [];
+
+        if (name instanceof Array)
+            reporters = name.map(r => typeof r === 'string' || typeof r === 'function' ? { name: r } : r);
+        else {
+            const reporter = { name, outStream };
+
+            reporters.push(reporter);
+        }
+
+        reporters.forEach(r => Runner._validateReporterOutput(r.outStream));
+
+        return reporters;
+    }
+
     // API
     embeddingOptions (opts) {
-        this._registerAssets(opts.assets);
-        this.opts.TestRunCtor = opts.TestRunCtor;
+        const { assets, TestRunCtor } = opts;
+
+        this._registerAssets(assets);
+        this.configuration.mergeOptions({ TestRunCtor });
 
         return this;
     }
 
     src (...sources) {
-        this.bootstrapper.sources = this.bootstrapper.sources.concat(flatten(sources));
+        if (this.apiMethodWasCalled.src)
+            throw new GeneralError(MESSAGE.multipleAPIMethodCallForbidden, OPTION_NAMES.src);
+
+        sources = flatten(sources);
+        this.configuration.mergeOptions({ [OPTION_NAMES.src]: sources });
+
+        this.apiMethodWasCalled.src = true;
 
         return this;
     }
 
     browsers (...browsers) {
-        this.bootstrapper.browsers = this.bootstrapper.browsers.concat(flatten(browsers));
+        if (this.apiMethodWasCalled.browsers)
```

Comment:
Perhaps we could refactor `FlagList` to throw an exception in a field setter?  Thus we won't need all the `if(this.apiMethodWasCalled)` conditions.

Reply:
I disagree with you. `FlagList` is a simple concept. Don't need to extend it with error generation.